tensorflow/compiler/mlir/tosa/transforms/legalize_tfl_stateful.cc

```diff
@@ -0,0 +1,208 @@
+/* Copyright 2023 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// Legalize TensorFlow Lite StatefulOps to TOSA
+
+#include <cstddef>
+#include <iostream>
+
+#include "llvm/ADT/ArrayRef.h"
+#include "mlir/Dialect/Quant/QuantTypes.h"  // from @llvm-project
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"  // from @llvm-project
+#include "mlir/Dialect/Traits.h"  // from @llvm-project
+#include "mlir/IR/BuiltinAttributes.h"  // from @llvm-project
+#include "mlir/IR/BuiltinTypes.h"  // from @llvm-project
+#include "mlir/IR/Matchers.h"  // from @llvm-project
+#include "mlir/IR/PatternMatch.h"  // from @llvm-project
+#include "mlir/Support/LLVM.h"  // from @llvm-project
+#include "mlir/Transforms/DialectConversion.h"  // from @llvm-project
+#include "tensorflow/compiler/mlir/lite/ir/tfl_ops.h"
+#include "tensorflow/compiler/mlir/tosa/transforms/passes.h"
+
+#define PASS_NAME "tosa-legalize-tfl-stateful"
+
+namespace mlir {
+namespace tosa {
+namespace {
+
+#define GEN_PASS_DEF_TOSALEGALIZETFLSTATEFULPASS
+#include "tensorflow/compiler/mlir/tosa/transforms/passes.h.inc"
+
+// Performs lowering tfl stateful operators to TOSA
+class TosaLegalizeTFLStateful : public impl::TosaLegalizeTFLStatefulPassBase<TosaLegalizeTFLStateful> {
+ public:
+  explicit TosaLegalizeTFLStateful() {}
+  void runOnOperation() override;
+};
+
+void TosaLegalizeTFLStateful::runOnOperation() {
+    auto* context = &getContext();
+    auto moduleOp = getOperation();
+    mlir::OpBuilder builder(moduleOp.getBodyRegion());
+
+    DenseMap<StringRef, func::FuncOp> symNameToFunction;
+    for (auto func : moduleOp.getOps<func::FuncOp>()) {
+      symNameToFunction[func.getSymName()] = func;
+    }
+
+    DenseMap<StringRef, DenseElementsAttr> sharedNameToConstant;
+    DenseMap<StringRef, LocationAttr> sharedNameToLoc;
+
+    llvm::SmallVector<mlir::TFL::VarHandleOp, 6> handleOps;
+    llvm::SmallVector<mlir::TFL::AssignVariableOp, 6> assignOps;
+    llvm::SmallVector<mlir::TFL::ReadVariableOp, 6> readOps;
+
+    for (auto it : symNameToFunction) {
+      auto func = std::get<1>(it);
+      // Look through the initialization functions and find the assigned values
+      // for each handle, save out the constant value.
+      for (auto init : func.getOps<mlir::TFL::CallOnceOp>()) {
```

Comment:
You have multiple `getOps` and finders nested within this assignment. You should likely iterate only once for each case and place them in a map rather than nesting. This could explode in computational complexity as the number of variables increase.

Reply:
done in the latest patch